firebase-database/src/main/java/com/google/firebase/database/core/Repo.java

```diff
@@ -972,6 +1008,10 @@ public void run() {
     }
   }
 
+  public <TResult> Task<TResult> runTransaction(Function<com.google.firebase.database.core.Transaction, Task<TResult>> updateFunction) {
```

Comment:
How should we queue transactions?

Currently, I'm thinking about having something like Firestore's `TransactionRunner` hold a reference to some queue. I'm not sure how that queue should look though.

Existing RTDB transactions have a tree of queues, each path has a queue that is a `List<TransactionData>`.

Transactions ready to be sent to the backend (in state `RUN`) are aggregated and sent in batches in `[sendReadyTransactions](https://github.com/firebase/firebase-android-sdk/blob/0de237d26e0336e83c8a494dec9aea9c8a69eece/firebase-database/src/main/java/com/google/firebase/database/core/Repo.java#L997)`.

The actual `compareAndPut` network request is done [here](https://github.com/firebase/firebase-android-sdk/blob/0de237d26e0336e83c8a494dec9aea9c8a69eece/firebase-database/src/main/java/com/google/firebase/database/core/Repo.java#L1055).  

All of the transactions at a location are batched into a single update, where the client figures out what hash of that location was before any updates, and then constructs a datasnapshot node using `[updateChild](https://github.com/firebase/firebase-android-sdk/blob/0de237d26e0336e83c8a494dec9aea9c8a69eece/firebase-database/src/main/java/com/google/firebase/database/core/Repo.java#L1047)`.

With these new transactions, it doesn't make sense to have a path-based queue. We still want a queue because transactions can be retried, but a `List<Transaction>` might be sufficient.